File: beam-series-python/model_guide/ptransform_group_ops.py
import logging
import apache_beam as beam
from apache_beam.options.pipeline_options import PipelineOptions

employees = [
        {
            "id": 0,
            "name": "John C",
            "joining_date": "2000-01-01",
            "dept_id": "D101",
            "is_active": True
            },
        {
            "id": 1,
            "name": "Tom D",
            "joining_date": "2002-02-01",
            "dept_id": "D102",
            "is_active": True
            },
        {
            "id": 2,
            "name": "Max",
            "joining_date": "2003-04-01",
            "dept_id": "D104",
            "is_active": False
            },
        {
            "id": 3,
            "name": "Bruce",
            "joining_date": "2003-05-01",
            "dept_id": "D102",
            "is_active": False
            },
        {
            "id": 4,
            "name": "Barry",
            "joining_date": "2003-04-22",
            "dept_id": "D101",
            "is_active": False
            },
        {
            "id": 5,
            "name": "Clark",
            "joining_date": "2001-04-01",
            "dept_id": "D103",
            "is_active": False
            },
        {
            "id": 6,
            "name": "Bryan C.",
            "joining_date": "2010-07-01",
            "dept_id": "D104",
            "is_active": True
            }
        ]

dept = [
        {
            "id": "D101",
            "dept_name": "Support"
            },
        {
            "id": "D102",
            "dept_name": "HR"
            },
        {
            "id": "D103",
            "dept_name": "Marketing"
            },
        {
            "id": "D104",
            "dept_name": "Sells"
            }
        ]

def flat_join(element: tuple):
    data = element[1]
    dept = data["dept"][0]["dept_name"]
    return [ (dept,employee) for employee in data["employees"]]

def run():
    pipeline_options = PipelineOptions()
    with beam.Pipeline(options=pipeline_options) as p:
        employees_pcoll = p | "Create employees pcoll" >> beam.Create(employees)
        dept_pcoll = p | "Create dept pcoll" >> beam.Create(dept)
        
        employees_kv = employees_pcoll | "Add keys for employees" >> beam.WithKeys(lambda x: x["dept_id"])
        dept_kv = dept_pcoll | "Add keys for dept" >> beam.WithKeys(lambda x: x["id"])

        joined_cogroup = ({"employees": employees_kv,"dept": dept_kv}) | "Join with CoGroupByKey" >> beam.CoGroupByKey()
        
        flatten_joined = joined_cogroup | "Flatten ops" >> beam.FlatMap(flat_join)

        min_joined_date_by_dept = (flatten_joined |
            "Group by dept" >> beam.GroupByKey() |
            "Take min of joining date" >> beam.Map(lambda x:
                (x[0],min([employee["joining_date"] for employee in x[1]]))
            )
        )

        min_joined_date_by_dept_2 = (flatten_joined |
            "Min of joining date Group by dept" >> beam.GroupBy(lambda x: x[0]).aggregate_field(
                lambda x: x[1]["joining_date"],min,"min_joining_date"
            )
        )

        min_joined_date_by_dept_2 | "Print debugger" >> beam.Map(print)
if __name__ == "__main__":
    logging.getLogger().setLevel(logging.INFO)
    run()

File: beam-series-go/model_guide/ptransform_group_ops.go
package main

import (
	"context"
	"log"

	"github.com/apache/beam/sdks/v2/go/pkg/beam"
	"github.com/apache/beam/sdks/v2/go/pkg/beam/register"
	"github.com/apache/beam/sdks/v2/go/pkg/beam/x/beamx"
	"github.com/apache/beam/sdks/v2/go/pkg/beam/x/debug"
)

type Employee struct {
  Id int
  Name string 
  JoiningDate string 
  DeptID string 
  IsActive bool 
}

type Dept struct {
  Id string
  DeptName string
}

func FlatJoin(key string, employeeIter func(*Employee) bool,deptIter func(*Dept) bool, emit func(string,Employee)) {
  var dept Dept
  var dept_name string
  for deptIter(&dept) {
    dept_name = dept.DeptName
  }

  var employee Employee
  for employeeIter(&employee) {
    emit(dept_name,employee)
  }
}

func init() {
  register.Function4x0(FlatJoin)
  register.Emitter2[string,Employee]()
  register.Iter1[Employee]()
  register.Iter1[Dept]()
}

func main() {

  employees := []Employee{
    Employee{
      0,
      "John C",
      "2000-01-01",
      "D101",
      true,
    },
    Employee{
      1,
      "Tom D",
      "2002-02-01",
      "D102",
      true,
    },
    Employee{
      2,
      "Max",
      "2003-04-01",
      "D104",
      false,
    },
    Employee{
      3,
      "Bruce",
      "2003-05-01",
      "D102",
      false,
    },
    Employee{
      4,
      "Barry",
      "2003-04-22",
      "D101",
      false,
    },
    Employee{
      5,
      "Clark",
      "2001-04-01",
      "D103",
      false,
    },
    Employee{
      6,
      "Bryan C.",
      "2010-07-01",
      "D104",
      true,
    },
  }
     
  dept := []Dept{
      Dept{
        "D101",
        "Support",
      },
      Dept{
        "D102",
        "HR",
      },
      Dept{
        "D103",
        "Marketing",
      },
      Dept{
        "D104",
        "Sells",
      },
  }

  beam.Init()

  p := beam.NewPipeline()

  s := p.Root()



  employees_pcoll := beam.CreateList(s,employees)
  dept_pcoll := beam.CreateList(s,dept)

  employees_kv := beam.ParDo(s, func(element Employee) (string,Employee) {
    return element.DeptID,element
  },employees_pcoll)
  
  dept_kv := beam.ParDo(s, func(element Dept) (string,Dept) {
    return element.Id,element
  },dept_pcoll)

  joined_cogroup := beam.CoGroupByKey(s,employees_kv,dept_kv)

  flatten_joined := beam.ParDo(s,FlatJoin,joined_cogroup)

  group_by_dept := beam.GroupByKey(s,flatten_joined)

  min_joining_date_per_dept := beam.ParDo(s, func(key string, element func(*Employee) bool) (string,string) {
      min_date := "9999-99-99"
      var employee Employee
      for element(&employee) {
        if (employee.JoiningDate <= min_date) {
          min_date = employee.JoiningDate
        }
      }
      return key, min_date
  },group_by_dept)

   
  debug.Print(s,min_joining_date_per_dept)

  if err := beamx.Run(context.Background(), p); err != nil {
    log.Fatalf("Failed to execute job: %v",err)
  }
}
